clients/src/main/java/org/apache/kafka/common/utils/SecurityUtils.java

```diff
@@ -34,4 +43,26 @@ public static KafkaPrincipal parseKafkaPrincipal(String str) {
         return new KafkaPrincipal(split[0], split[1]);
     }
 
+    public static void addConfiguredSecurityProviders(Map<String, ?> configs) {
+        String securityProviderClassesStr = (String) configs.get(SecurityConfig.SECURITY_PROVIDERS_CONFIG);
+        if (securityProviderClassesStr == null || securityProviderClassesStr.equals("")) {
+            return;
+        }
+        try {
+            String[] securityProviderClasses = securityProviderClassesStr.replaceAll("\\s+", "").split(",");
+            for (int index = 0; index < securityProviderClasses.length; index++) {
+                SecurityProviderCreator securityProviderCreator = (SecurityProviderCreator) Class.forName(securityProviderClasses[index]).newInstance();
+                securityProviderCreator.configure(configs);
+                Security.insertProviderAt(securityProviderCreator.getProvider(), index + 1);
+            }
+        } catch (ClassCastException e) {
+            LOGGER.error("Generator provided through " + SecurityConfig.SECURITY_PROVIDERS_CONFIG +
```

Comment:
Change Generator to Creator?

Reply:
Here we rely on insertProviderAt() programatic way BUT if in the application's context somebody else calls Security.insertProviderAt(provider,1) that provider will be given the priority for any conflicting Provider services+algorithms. This code works well if you have exclusive services+algorithms example SPIFFE but if you are writing a provider for Standard algorithms example TrustManagerFactory.PKIX then you may run into trouble since your insertProviderAt() call got overridden by somebody else in the application context/startup.

When that happens I don't know easy way to fix it. I think It is important to call this out.

Reply:
We have this kind of challenges today in some other context. Hence speaking from the real experience! :)

Reply:
Set provider only at the beginning of loading application per JVM will avoid what you said.

Reply:
So let us say - we already have a single provider for JSSE but for Kafka we need one more to override with some of the common services+algorithm (example only override TrustManagerFactory.PKIX). We are at the mercy of the initialization sequence of calls, isn't it? Here the different providers could be owned by different teams and when we are in bigger infra setup it may be difficult to overcome this technical limitation. Only for Kafka applications the init sequence is different vs rest of the infra in the company.

Basically, we are making an assumption which may not hold true and then we will be really stuck. What I am suggesting is - If we are calling out that limitation with these changes it is okay but otherwise it will result in a bug.

Reply:
I think, for most scenarios, using a single provider for JSSE per JVM should be fine.  At some cases, If need specific providers for some components, then the component self need explicitly specify the name of the provider such as:
public static final KeyManagerFactory getInstance(String algorithm, String provider)?

Reply:
That has its own complications because if my provider is only providing TrustManagerFacotry.PKIX then it can't provider other services+algorithms that might be expected in calls like SSLContext.getInstance(String protocol, String provider). SSLContext.getInstance might look for TLSv1.1 or TLSv1.2 etc which my provider doesn't really have and I don't have a way to fallback anymore once I go route of specifying "provider" in getInstance() calls.